docs(api): Update publishing docs (#1033)
getsentry/develop: src/docs/api/public.mdx

don't control the version of the client a user runs.

### External Client

An external client is a client that Sentry does not control; for example, a script written by a
customer run in their repository, or an integration written using the Integration Platform. For this
type of client, we have no control over the endpoints they choose to use, nor can we require that
they immediately update their endpoint.

### Private API

A private API is an undocumented API. Sentry can make breaking changes to this type of API, and no
one will blink an eye.

### Public API

Our public API is documented at [https://docs.sentry.io/api/](https://docs.sentry.io/api/).

### Private Attributes in a Public Endpoint

You can have private attributes within a public endpoint.

As an example:
[https://docs.sentry.io/api/teams/retrieve-a-team/](https://docs.sentry.io/api/teams/retrieve-a-team/). 
The response has multiple attributes:

```json
{
  "id": "2",
  "slug": "the-interstellar-jurisdiction"
  ...
}
```

Let's say we also return a `nickname` for a team. If what is returned is not documented as a
response in our public documentation, that attribute is a private attribute within a public
endpoint.

> ### DRF (Django REST Framework)

Author: Added this to a different page here: https://github.com/getsentry/develop/pull/1032
I don't think it's related to publishing

## When to make an API public?

> Should I publish my API? The answer in most of the cases is **Yes**. As a developer facing company it is critical to make Sentry features available for developers to use in their various automations. Do not try to predict if an automation finds the API useful because we can't predict all the automations our clients might want to build. With that being said, there are 3 exceptions where we don't publish APIs:

Reviewer: ```suggestion
Should I publish my API? The answer in most cases is **Yes**. As a developer facing company it is critical to make Sentry features available for developers to use in their various automations. Do not try to predict if an automation finds the API useful because we can't predict all the automations our clients might want to build. With that being said, there are 3 exceptions where we don't publish APIs:
```

1. **Experimental APIs:** APIs that are not fully launched or will go through multiple iterations in a short time that could cause breaking changes. These are expected to be published or removed after a while.
2. **Private APIs:** APIs that are controlling a UI feature in Sentry, for example if a button is visible. 
3. **APIs in getsentry:** If you're a Sentry engineer and building APIs in getsentry repo, know that they won't be published. This is the expected behavior for all the getsentry APIs, but if you have a case that you think it should be published let [owners-api](https://github.com/orgs/getsentry/teams/owners-api) know and we can help with that.

> **Note**: Always keep in mind that your APIs will be used out of Sentry product even if they're `private`. All the APIs in Sentry repo are accessible to public due to the open-source nature.  

Reviewer: ```suggestion
**Note**: Always keep in mind that your APIs will be used out of Sentry product even if they're `private`. All the APIs in Sentry repo are accessible to public due to our open-source nature.  
```

### Before Publishing

> Remember, making an API public means that once an endpoint is public, you can only add to it: You cannot make _any_ breaking

Reviewer: ```suggestion
Remember, making an API public means that you can only add to it: You cannot make _any_ breaking
```

@extend_schema(tags=["Teams"])
class OrganizationTeamsEndpoint(...):
    owner = ApiOwner.ENTERPRISE
    publish_status = {
        'GET': ApiPublishStatus.PUBLIC, 
        'POST': ApiPublishStatus.PUBLIC,
        'PUT': ApiPublishStatus.EXPERIMENTAL,
    }
```
### 4. Endpoint Description

Specify the endpoint description in the documentation using the endpoint's docstring. Please make sure to include a description for the resources involved as well if they are not clear. For example, `team` and `organization` here don't require

> description but other words like `external-issue` or `integration` are vauge and need description. 

Reviewer: ```suggestion
description but other words like `external-issue` or `integration` are vague and need description. 
```

        """
```

Here's description of each field in the method decorator:
- **operation_id:** will be shown as the title of the endpoint's page in the
documentation.

> - **parameters:** is a list of path and query parameters and if they're required. 

Reviewer: ```suggestion
- **parameters:** is a list of path and query parameters, and whether or not they are required. 
```

  - You can find existing parameters in this 
  [file](https://github.com/getsentry/sentry/blob/master/src/sentry/apidocs/parameters.py). Note
  that the `description` field in `OpenApiParameter` populates the parameter's description in the
  documentation.
  - DRF serializers conveniently translate into query parameters. See
  [here](https://github.com/getsentry/sentry/blob/0b9b6563bc4e232334cfc71f136a49117171d13f/src/sentry/api/endpoints/organization_stats_v2.py#L145)
  for an example of this.

> - **request:** is the request serializer that can generally just be the DRF serializer of endpoint itself. If you need more

Reviewer: ```suggestion
- **request:** is the request serializer that can generally just be the DRF serializer of the endpoint itself. If you need more
```

},
```

### Private Attributes in a Public Endpoint

You can have private attributes within a public endpoint.

As an example:
[https://docs.sentry.io/api/teams/retrieve-a-team/](https://docs.sentry.io/api/teams/retrieve-a-team/). 
The response has multiple attributes:

```json
{
  "id": "2",
  "slug": "the-interstellar-jurisdiction"
  ...
}
```

> Let's say we also return a `nickname` for a team. If what is returned is not documented as a
> response in our public documentation, that attribute is a private attribute within a public
> endpoint.

Reviewer 2: ```suggestion
Let's say we also return a `nickname` for a team. If the data returned is not documented as a
response in our public documentation, that attribute is a private attribute within a public
endpoint.
```

Once you make changes to an endpoint and merge the change into Sentry, a series of GitHub Actions
will be triggered to make your changes automatically go live:

1. The [openapi workflow in
   sentry](https://github.com/getsentry/sentry/blob/master/.github/workflows/openapi.yml) updates
   the schema in [sentry-api-schema](https://github.com/getsentry/sentry-api-schema) with the
   OpenAPI build artifact.
2. The [cascade-to-sentry-docs workflow in
   sentry-api-schema](https://github.com/getsentry/sentry-api-schema/blob/main/.github/workflows/cascade-to-sentry-docs.yml)
   reacts to the push to `main` in (1) by triggering the [bump-api-schema-sha workflow in
   sentry-docs](https://github.com/getsentry/sentry-docs/blob/master/.github/workflows/bump-api-schema-sha.yml).
3. The [bump-api-schema-sha workflow in
   sentry-docs](https://github.com/getsentry/sentry-docs/blob/master/.github/workflows/bump-api-schema-sha.yml)
   fetches the latest commit SHA from `sentry-api-schema` and writes it into the correct file, then
   makes and merges a PR in `sentry-docs`, which kicks off a deploy via Vercel to
   https://docs.sentry.io/api/.

## Requesting an API to be public

> Are you a Sentry client who wants an endpoint to be public?

Reviewer 2: ```suggestion
Are you a Sentry user who wants an endpoint to be public?
```
